File: Year 1/Term 2/Programming/C++/Lab 5/Lab_5/Lab_5/Point.cpp
#include "Point.h"

//constructors:
Point::Point(): m_x(0), m_y(0){}
Point::Point(float x, float y) : m_x(x), m_y(y) {}
Point::Point(const Point& other): m_x(other.m_x), m_y(other.m_y){}

//methods:
void Point::set(float x, float y) { m_x = x; m_y = y; }
void Point::set_x(float x) { m_x = x;}
void Point::set_y(float y) { m_y = y; }
int Point::get_x() const { return m_x; }
int Point::get_y() const { return m_y; }

//operators:
Point Point::operator+(const Point& other) {
	Point tmp;
	tmp.m_x = m_x + other.m_x;
	tmp.m_y = m_y + other.m_y;
	return tmp;
}
void Point::operator+=(const Point& other){
	m_x += other.m_x;
	m_y += other.m_y;
}
void Point::operator/=(const int number) {
	m_x /= number;
	m_y /= number;
}
//friend functions
ostream& operator <<(ostream& out, const Point& point) {
	out << "(" << point.m_x << "; " << point.m_y << ")";
	return out;
}
float calc_distance(const Point& point_a, const Point& point_b) {
	float x = pow(point_a.m_x - point_b.m_x, 2);
	float y = pow(point_a.m_y - point_b.m_y, 2);
	return sqrt(x+y);
}

File: Year 1/Term 2/Programming/Python/Lab 5/pythonProject/Point.py
from math import sqrt
from accessify import private, protected


class Point:
    def __new__(cls, *args, **kwargs):
        return super().__new__(cls)

    def __init__(self, x: float = 0, y: float = 0):
        self.__x = x
        self.__y = y

    def display(self):
        print("(" + str(self.__x) + "; " + str(self.__y) + ")")

    def set_x(self, x):
        if type(x) in (int, float):
            self.__x = x

    def set_y(self, y):
        if type(y) in (int, float):
            self.__y = y

    def set(self, x, y):
        if type(x) in (int, float) and type(y) in (int, float):
            self.__x = x
            self.__y = y

    def get_x(self):
        return self.__x

    def get_y(self):
        return self.__y

    def __add__(self, other):
        if type(other) == Point:
            x = self.__x + other.__x
            y = self.__y + other.__y
            return Point(x, y)

    def __increase__(self, other):
        if type(other) == Point:
            self.__x += other.__x
            self.__y += other.__y
        return

    def __div_avg__(self, number):
        if type(number) == int:
            self.__x /= number
            self.__y /= number
        return

    @staticmethod
    def calc_distance(point_a, point_b):
        return sqrt((point_a.__x - point_b.__x)**2 + (point_a.__y - point_b.__y)**2)

    @staticmethod
    def calc_sq_triangle(point_a, point_b, point_c):
        side_a = Point.calc_distance(point_a, point_b)
        side_b = Point.calc_distance(point_b, point_c)
        side_c = Point.calc_distance(point_c, point_a)
        semi_p = (side_a + side_b + side_c) / 2
        return sqrt(semi_p * (semi_p - side_a) * (semi_p - side_b) * (semi_p - side_c))

    @staticmethod
    def calc_avg(num, points):
        tmp = Point()
        for i in range(num):
            tmp.__increase__(points[i])
        tmp.__div_avg__(num)
        return tmp
